### ISSUE TITLE
App Lock

### ISSUE THREAD

**Issue by Valodim** [open-keychain/open-keychain#2163]
## Summary

We would like to move away from individually encrypted keys, towards an app lock mechanism. It's too difficult to create an actual migration, so we will try to break this down into steps that we can deploy incrementally.

## Plan

The ground work for this has been laid in #2066, but it's still missing quite a few things. Based on a few discussions, here is the rough plan we have in mind:

- [ ] Add app lock to settings. This is only available when all keys have no password, which should be the default case for installations that don't already have an encrypted key from a previous version. This is the preferred way to protect keys.
- [ ] Move "change password" option from overflow menu in key display into a tab under advanced.
- [ ] Remove "choose password" from key creation.
- [ ] Decrypt keys in symmetrically encrypted backup.
- [ ] Ask for key passphrase during import, and decrypt.
- [ ] (low prio) Hide individual passwords for new installations.

## Out of scope (for now)

- For now, we won't encrypt the entire database. The most important reason is that this would break the default workflow in K-9 of showing recipient information.
- A full migration is very difficult, and bugs could result in loss of secret keys. We will need to keep support for individual passphrases

**Comment by dllud** [comment]
Will it be possible to create/import new keys and use the old password encryption scheme (even if under some advanced setting)? Or will the multiple-key/multiple-passphrase become impossible to achieve for all new keys?

**Comment by Valodim** [comment]
Not sure yet. An app lock mechanism is a much easier mental model that users can easily understand, which is why we want to demote per-key passphrases to a supported (rather than encouraged) workflow. Dropping it entirely is an option that we will consider, but not anytime soon.

**Comment by zeapo** [comment]
If I understand well. You want to self-encrypt the passphrase for each key added to OpenKeychain using a per-app passphrase (would be like a master password for each app). Or am I misunderstanding completely?

**Comment by dllud** [comment]
> would be like a master password for each app

Nope. It will be just a single master password for the entire OpenKeychain app.
(The only per-app setting that you have, and that's already implement, is the mechanism to restrict which keys a given app can access).
Discussion should not elongate here, there is an issue for that: #1541

**Comment by Valodim** [comment]
Actually, this is the place to discuss now. Way too much text in the other one that is partly outdated, and noone is going to read anyways :)

**Comment by fabianski7** [comment]
any progress?
